src/firejail/preproc.c

```diff
@@ -21,12 +21,61 @@
 #include <sys/mount.h>
 #include <sys/stat.h>
 #include <sys/types.h>
+#include <sys/file.h>
 #include <dirent.h>
+#include <fcntl.h>
 
 static int tmpfs_mounted = 0;
 
+static void preproc_lock_file(const char *path, int *lockfd_ptr) {
+	if( *lockfd_ptr == -1 ) {
+		if (arg_debug)
+			fprintf(stderr, "pid=%d: lock %s ...\n", (int)getpid(), path);
+		int lockfd = open(path, O_WRONLY | O_CREAT | O_CLOEXEC, S_IRUSR | S_IWUSR);
+		if( lockfd == -1 )
+			errExit(path);
+
+		if (fchown(lockfd, 0, 0) == -1)
+			fprintf(stderr, "Error: cannot chown root:root %s: %m\n", path);
+
+		if (flock(lockfd, LOCK_EX) == -1) {
+			fprintf(stderr, "Error: cannot lock %s: %m\n", path);
+			close(lockfd);
+		}
+		else {
+			if (arg_debug)
+				fprintf(stderr, "pid=%d: locked %s\n", (int)getpid(), path);
+			*lockfd_ptr = lockfd;
+		}
+	}
+}
+
+static void preproc_unlock_file(const char *path, int *lockfd_ptr) {
+	int lockfd = *lockfd_ptr;
+	if (lockfd != -1) {
+		*lockfd_ptr = -1;
+		if (arg_debug)
+			fprintf(stderr, "pid=%d: unlock %s\n", (int)getpid(), path);
+		if (flock(lockfd, LOCK_UN) == -1)
+			fprintf(stderr, "Error: cannot unlock %s: %m\n", path);
+		if (close(lockfd) == -1)
+			fprintf(stderr, "Error: cannot close %s: %m\n", path);
+	}
+}
```

Comment:
```suggestion
static void preproc_lock_file(const char *path, int *lockfd_ptr) {
	assert(path != NULL);
	assert(lockfd_ptr != NULL);

	// Already locked
	if (*lockfd_ptr != -1)
		return;

	long pid = (long)getpid();
	if (arg_debug)
		fprintf(stderr, "pid=%ld: locking %s ...\n", pid, path);

	int lockfd = open(path, O_WRONLY | O_CREAT | O_CLOEXEC, S_IRUSR | S_IWUSR);
	if (lockfd == -1) {
		fprintf(stderr, "Error: cannot create a lockfile at %s\n", path);
		errExit("open");
	}

	if (fchown(lockfd, 0, 0) == -1) {
		fprintf(stderr, "Error: cannot chown root:root %s\n", path);
		errExit("fchown");
	}

	if (flock(lockfd, LOCK_EX) == -1) {
		fprintf(stderr, "Error: cannot lock %s\n", path);
		errExit("flock");
	}

	*lockfd_ptr = lockfd;
	if (arg_debug)
		fprintf(stderr, "pid=%ld: locked %s\n", pid, path);
}

static void preproc_unlock_file(const char *path, int *lockfd_ptr) {
	assert(path != NULL);
	assert(lockfd_ptr != NULL);

	int lockfd = *lockfd_ptr;
	// Already unlocked
	if (lockfd == -1)
		return;

	long pid = (long)getpid();
	if (arg_debug)
		fprintf(stderr, "pid=%ld: unlocking %s ...\n", pid, path);

	if (flock(lockfd, LOCK_UN) == -1) {
		fprintf(stderr, "Error: cannot unlock %s\n", path);
		errExit("flock");
	}

	if (close(lockfd) == -1) {
		fprintf(stderr, "Error: cannot close %s\n", path);
		errExit("close");
	}

	*lockfd_ptr = -1;
	if (arg_debug)
		fprintf(stderr, "pid=%ld: unlocked %s\n", pid, path);
}
```

Main changes:

* Fix formatting
* Reduce indentation/return early
* Add asserts
* Avoid non-standard `%m`
* Make messages more consistent
* Abort if fchown fails
* Abort on most/all errors

Since all flock operations are synchronous/blocking in this code, is there a
reason not to abort if they fail?

Reply:
Well, for the most part this just moves original behavior to helper functions. I did add one errExit, but the original flock() code ignored errors altogether -> suggests that limping forwards is preferred over bailout.

I'll modify the pr to address these points

Reply:
... or rather, picked suggested changes as-is. My first time bumping into this feature. Thanks,

Reply:
> Well, for the most part this just moves original behavior to helper
> functions.

> I did add one errExit

> ```diff
>  	int lockfd = open(path, O_WRONLY | O_CREAT | O_CLOEXEC, S_IRUSR | S_IWUSR);
> -		if( lockfd == -1 )
> -			errExit(path);
> +	if (lockfd == -1) {
> +		fprintf(stderr, "Error: cannot create a lockfile at %s\n", path);
> +		errExit("open");
> +	}
> ```

This was already changed, but to clarify, when calling errExit please treat the
first argument as if it were the format string in printf (that is, prefer using
a string literal), as this is how it is usually used and eventually it should
support variable arguments like printf.

> but the original flock() code ignored errors altogether -> suggests that
> limping forwards is preferred over bailout.

True, but from what I understand, valid errors in those calls in this code are
so unlikely (like ENOLCK in flock) that if any error happens it's likely that
something went very wrong anyway.

Though admittedly I'm not sure about signal interruption.